### Amazon SageMaker Asynchronous Inference with Hugging Face Model
_**A new near real-time Inference option for generating machine learning model predictions. This notbook require Python3(ipykernel) with SageMaker >= 2.75**_

**Table of Contents**

* [Background](#background)
* [Notebook Scope](#scope)
* [Overview and sample end to end flow](#overview)
* [Section 1 - Setup](#setup) 
    * [Create Model](#createmodel)
    * [Create EndpointConfig](#endpoint-config)
    * [Create Endpoint](#create-endpoint)
* [Section 2 - Using the Endpoint](#endpoint) 
    * [Invoke Endpoint](#invoke-endpoint)
    * [Check Output Location](#check-output) 
* [Section 3 - Clean up](#clean)

### Background <a id='background'></a>  
Amazon SageMaker Asynchronous Inference is a new capability in SageMaker that queues incoming requests and processes them asynchronously. SageMaker currently offers two inference options for customers to deploy machine learning models: 1) a real-time option for low-latency workloads 2) Batch transform, an offline option to process inference requests on batches of data available upfront. Real-time inference is suited for workloads with payload sizes of less than 6 MB and require inference requests to be processed within 60 seconds. Batch transform is suitable for offline inference on batches of data. 

Asynchronous inference is a new inference option for near real-time inference needs. Requests can take up to 15 minutes to process and have payload sizes of up to 1 GB. Asynchronous inference is suitable for workloads that do not have sub-second latency requirements and have relaxed latency requirements. For example, you might need to process an inference on a large image of several MBs within 5 minutes. In addition, asynchronous inference endpoints let you control costs by scaling down endpoints instance count to zero when they are idle, so you only pay when your endpoints are processing requests. 

### Notebook scope <a id='scope'></a>  
This notebook provides an introduction on how to use the SageMaker Asynchronous inference capability with Hugging Face models. This notebook will cover the steps required to create an Asynchronous inference endpoint and test it with some sample requests. 

### Overview <a id='overview'></a>
Asynchronous inference endpoints have many similarities (and some key differences) compared to real-time endpoints. The process to create asynchronous endpoints is similar to real-time endpoints. You need to create: a model, an endpoint configuration, and then an endpoint. However, there are specific configuration parameters specific to asynchronous inference endpoints which we will explore below. 

Invocation of asynchronous endpoints differ from real-time endpoints. Rather than pass request payload inline with the request, you upload the payload to Amazon S3 and pass an Amazon S3 URI as a part of the request. Upon receiving the request, SageMaker provides you with a token with the output location where the result will be placed once processed. Internally, SageMaker maintains a queue with these requests and processes them. During endpoint creation, you can optionally specify an Amazon SNS topic to receive success or error notifications. Once you receive the notification that your inference request has been successfully processed, you can access the result in the output Amazon S3 location. 

---
## 1. Setup <a id='setup'></a>

First we ensure we have an updated version of Sagemaker, which includes the latest SageMaker features:

Import the required python libraries:

In [1]:
!python -m pip install --upgrade pip --quiet
!pip install -U awscli --quiet
!pip install --upgrade sagemaker --quiet

In [1]:
from time import gmtime, strftime
from sagemaker import image_uris
import sagemaker
import logging
import boto3
import json

In [2]:
logger = logging.getLogger("__name__")
logger.setLevel(logging.DEBUG)
logger.addHandler(logging.StreamHandler())

In [3]:
logger.info(f"Using SageMaker version: {sagemaker.__version__}")

Using SageMaker version: 2.111.0


In [4]:
region = sagemaker.Session().boto_region_name
#role = sagemaker.get_execution_role()
role = 'arn:aws:iam::976939723775:role/service-role/AmazonSageMaker-ExecutionRole-20210317T133000'
boto3.setup_default_session(region_name=region)
boto_session = boto3.Session(region_name=region)
sm_session = sagemaker.session.Session()
sagemaker_client = boto_session.client("sagemaker")
sm_runtime = boto_session.client("sagemaker-runtime")
s3_bucket = sm_session.default_bucket()
current_timestamp = strftime("%m-%d-%H-%M", gmtime())
logger.info(f"Region = {region}")
logger.info(f"Role = {role}")

Region = us-west-2
Role = arn:aws:iam::976939723775:role/service-role/AmazonSageMaker-ExecutionRole-20210317T133000


Specify your IAM role. Go the AWS IAM console (https://console.aws.amazon.com/iam/home) and add the following policies to your IAM Role:
* SageMakerFullAccessPolicy
* Amazon S3 access: Apply this to get and put objects in your Amazon S3 bucket. Replace `bucket_name` with the name of your Amazon S3 bucket:      

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:AbortMultipartUpload",
                "s3:ListBucket"
            ],
            "Effect": "Allow",
            "Resource": "arn:aws:s3:::<bucket_name>/*"
        }
    ]
}
```

* (Optional) Amazon SNS access: Add `sns:Publish` on the topics you define. Apply this if you plan to use Amazon SNS to receive notifications.

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "sns:Publish"
            ],
            "Effect": "Allow",
            "Resource": "arn:aws:sns:us-east-2:123456789012:MyTopic"
        }
    ]
}
```

* (Optional) KMS decrypt, encrypt if your Amazon S3 bucket is encrypted.

Specify your SageMaker IAM Role (`role`) and Amazon S3 bucket . You can optionally use a default SageMaker Session IAM Role and Amazon S3 bucket. Make sure the role you use has the necessary permissions for SageMaker, Amazon S3, and optionally Amazon SNS.

### 1.1 Create Model <a id='createmodel'></a>
Specify the location of the pre-trained model stored in Amazon S3. This example uses a pre-trained Hugging Face model name (https://huggingface.co/finiteautomata/beto-sentiment-analysis) sentimentanalysis.tar.gz. The full Amazon S3 URI is stored in a string variable `MODEL_DATA_URL`. 

In [5]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained('finiteautomata/beto-sentiment-analysis')
tokenizer.save_pretrained('../../models/hf_hub/')

Downloading:   0%|          | 0.00/242k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/67.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/528 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/841 [00:00<?, ?B/s]

('../../models/hf_hub/tokenizer_config.json',
 '../../models/hf_hub/special_tokens_map.json',
 '../../models/hf_hub/vocab.txt',
 '../../models/hf_hub/added_tokens.json')

In [6]:
#!cd ../../models/hf_hub
#!pwd
#!tar zcvf sentimentanalysis.tar.gz *
#!cd -
s3_prefix = 'huggingface/hf_hub'
boto3.Session().resource('s3').Bucket(s3_bucket).Object(s3_prefix+'/sentimentanalysis.tar.gz').upload_file('../../models/hf_hub/sentimentanalysis.tar.gz')

In [7]:
MODEL_DATA_URL = f"s3://{s3_bucket}/{s3_prefix}/sentimentanalysis.tar.gz"

Specify a primary container. For the primary container, you specify the Docker image that contains inference code, artifacts (from prior training), and a custom environment map that the inference code uses when you deploy the model for predictions. In this example, we retrieve the appropriate container image by specifying the right framework version and framework details. Here in this case we are downloading container image associated with Hugging Face framework. For further details on right container images to use for your use case please refer to this link https://github.com/awsdocs/amazon-sagemaker-developer-guide/blob/master/doc_source/ and look in to appropriate ecr folder pertaining to the region of your interest

In [9]:
ecr_image = image_uris.retrieve(
    framework="huggingface",
    region=region,
    version="4.17.0",
    image_scope="inference",
    base_framework_version="pytorch1.10.2",
    py_version="py38",
    container_version="ubuntu20.04",
    instance_type="ml.m5.xlarge",
)
ecr_image

'763104351884.dkr.ecr.us-west-2.amazonaws.com/huggingface-pytorch-inference:1.10.2-transformers4.17.0-cpu-py38-ubuntu20.04'

Latest from https://github.com/aws/deep-learning-containers/blob/master/available_images.md
    
   Torch: 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-inference:1.12.1-gpu-py38-cu113-ubuntu20.04-sagemaker
   HF:  763104351884.dkr.ecr.us-east-1.amazonaws.com/huggingface-pytorch-inference:1.10.2-transformers4.17.0-gpu-py38-cu113-ubuntu20.04

In [12]:
model_name = f"beto-sentiment-analysis-async-1005"

Create a model by specifying the `ModelName`, the `ExecutionRoleARN` (the ARN of the IAM role that Amazon SageMaker can assume to access model artifacts/ docker images for deployment), and the `PrimaryContainer`.

In [13]:
response = sagemaker_client.create_model(
    ModelName=model_name,
    ExecutionRoleArn=role,
    PrimaryContainer={
        "Image": ecr_image,
        "ModelDataUrl": MODEL_DATA_URL,
        "Environment": {
            "HF_MODEL_ID": "finiteautomata/beto-sentiment-analysis",
            "HF_TASK": "text-classification",
            "SAGEMAKER_CONTAINER_LOG_LEVEL": "20",
            "SAGEMAKER_REGION": region,
        },
    },
)
model_arn = response["ModelArn"]

logger.info(f"Created Model: {model_arn}")

Created Model: arn:aws:sagemaker:us-west-2:976939723775:model/beto-sentiment-analysis-async-1005


In [14]:
endpoint_config_name = model_name

### 1.2 Create EndpointConfig <a id='endpointconfig'></a>

Once you have a model, create an endpoint configuration with CreateEndpointConfig. Amazon SageMaker hosting services uses this configuration to deploy models. In the configuration, you identify one or more models that were created using with CreateModel API, to deploy the resources that you want Amazon SageMaker to provision. Specify the AsyncInferenceConfig object and provide an output Amazon S3 location for OutputConfig. You can optionally specify Amazon SNS topics on which to send notifications about prediction results.

In [15]:
response = sagemaker_client.create_endpoint_config(
    EndpointConfigName=endpoint_config_name,
    ProductionVariants=[
        {
            "VariantName": "variant-1",
            "ModelName": model_name,
            "InstanceType": "ml.m5.xlarge",
            "InitialInstanceCount": 1,
        }
    ],

    AsyncInferenceConfig={
        "ClientConfig": { 
            "MaxConcurrentInvocationsPerInstance": 10
        },
        "OutputConfig": {
            "S3OutputPath": f"s3://{s3_bucket}/{s3_prefix}/output",
#            # Optionally specify Amazon SNS topics
#            # "NotificationConfig": {
#            #   "SuccessTopic": "arn:aws:sns:us-east-2:123456789012:MyTopic",
#            #   "ErrorTopic": "arn:aws:sns:us-east-2:123456789012:MyTopic",
#            # }
       },
    }
)
        
endpoint_config_arn = response["EndpointConfigArn"]
logger.info(f"Created EndpointConfig: {endpoint_config_arn}")

Created EndpointConfig: arn:aws:sagemaker:us-west-2:976939723775:endpoint-config/beto-sentiment-analysis-async-1005


### 1.3 Create Endpoint <a id='create-endpoint'></a>

Once you have your model and endpoint configuration, use the CreateEndpoint API to create your endpoint. The endpoint name must be unique within an AWS Region in your AWS account.

In [16]:
endpoint_name = model_name
response = sagemaker_client.create_endpoint(
    EndpointName="HuggingFaceAsyncEndpoint", EndpointConfigName="beto-sentiment-analysis-async"
)
endpoint_arn = response["EndpointArn"]
logger.info(f"Created Endpoint: {endpoint_arn}")

Created Endpoint: arn:aws:sagemaker:us-west-2:976939723775:endpoint/huggingfaceasyncendpoint


### Wait untill the endpoint has been created successfully

In [17]:
waiter = sagemaker_client.get_waiter("endpoint_in_service")
print("Waiting for endpoint to create...")
waiter.wait(EndpointName="HuggingFaceAsyncEndpoint")
resp = sagemaker_client.describe_endpoint(EndpointName="HuggingFaceAsyncEndpoint")
print(f"Endpoint Status: {resp['EndpointStatus']}")

Waiting for endpoint to create...
Endpoint Status: InService


--- 
## 2. Using the Endpoint <a id='endpoint'></a>

### 2.1 Uploading the Request Payload <a id='upload'></a>

Sample input.json placed in the input location

{"inputs": ["I like you. I love you","This is sad","am so happy that i want to cry","async endpoints are awesome"]}

In [18]:
input_s3_location = f"s3://{s3_bucket}/{s3_prefix}/input.json"
print(input_s3_location)

s3://sagemaker-us-west-2-976939723775/huggingface/hf_hub/input.json


### 2.1 Invoke Endpoint   <a id='invoke-endpoint'></a>

Get inferences from the model hosted at your asynchronous endpoint with InvokeEndpointAsync. Specify the location of your inference data in the InputLocation field and the name of your endpoint for EndpointName. The response payload contains the output Amazon S3 location where the result will be placed.

In [32]:
sagemaker_client.describe_endpoint(EndpointName="HuggingFaceAsyncEndpoint",)

{'EndpointName': 'HuggingFaceAsyncEndpoint',
 'EndpointArn': 'arn:aws:sagemaker:us-west-2:976939723775:endpoint/huggingfaceasyncendpoint',
 'EndpointConfigName': 'beto-sentiment-analysis-async',
 'ProductionVariants': [{'VariantName': 'variant-1',
   'DeployedImages': [{'SpecifiedImage': '763104351884.dkr.ecr.us-west-2.amazonaws.com/huggingface-pytorch-inference:1.7.1-transformers4.6.1-cpu-py36-ubuntu18.04',
     'ResolvedImage': '763104351884.dkr.ecr.us-west-2.amazonaws.com/huggingface-pytorch-inference@sha256:97cdf11484b82818b195579c7b5d8f16bc97d600ae352f47667e0587de7ae7f0',
     'ResolutionTime': datetime.datetime(2022, 10, 5, 21, 30, 26, 393000, tzinfo=tzlocal())}],
   'CurrentWeight': 1.0,
   'DesiredWeight': 1.0,
   'CurrentInstanceCount': 1,
   'DesiredInstanceCount': 1}],
 'EndpointStatus': 'InService',
 'CreationTime': datetime.datetime(2022, 10, 5, 21, 30, 24, 984000, tzinfo=tzlocal()),
 'LastModifiedTime': datetime.datetime(2022, 10, 5, 21, 32, 30, 26000, tzinfo=tzlocal()),


In [34]:
i=0
for i in range(5000):
    response = sm_runtime.invoke_endpoint_async(
        EndpointName="HuggingFaceAsyncEndpoint",
        InputLocation=input_s3_location,
        ContentType="application/json",
    )
    #print(response['OutputLocation'])

In [31]:
response['OutputLocation']

's3://sagemaker-us-west-2-976939723775/huggingface/hf_hub/output/cb22f063-73e6-4a7b-9cb2-b849c36cd33d.out'

### 2.2 Check Output Location <a id='check-output'></a>

Check the output location to see if the inference has been processed.

Sample inference output processed and  placed in the output location

[{"label":"POS","score":0.9982852339744568},{"label":"NEG","score":0.9333241581916809},{"label":"POS","score":0.595783531665802},{"label":"NEU","score":0.9964613318443298}]

### 3. Summary & Clean up <a id='clean'></a>

To Summarize, In this notebook we learned how to use the SageMaker Asynchronous inference capability with pre-trained Hugging Face models.

If you enabled auto-scaling for your endpoint, ensure you deregister the endpoint as a scalable target before deleting the endpoint. To do this, run the following:

In [22]:
response = sagemaker_client.deregister_scalable_target(
    ServiceNamespace="sagemaker",
    ResourceId="resource_id",
    ScalableDimension="sagemaker:variant:DesiredInstanceCount",
)

AttributeError: 'SageMaker' object has no attribute 'deregister_scalable_target'

Remember to delete your endpoint after use as you will be charged for the instances used in this Demo. 

You may also want to delete any other resources you might have created such as SNS topics, S3 objects, etc.